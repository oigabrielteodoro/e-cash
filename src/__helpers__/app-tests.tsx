import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
