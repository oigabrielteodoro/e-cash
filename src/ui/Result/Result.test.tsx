import React from 'react'

import { render, screen } from '__helpers__/app-tests'

import { Result } from '.'

function MockedComponent() {
  return (
    <Result status='success' title='Title' description='Description'>
      <button>Click me</button>
    </Result>
  )
}

describe('Result', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<MockedComponent />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(screen.getByText('Title')).toBeInTheDocument()
    expect(screen.getByText('Description')).toBeInTheDocument()
    expect(screen.getByText('Click me')).toBeInTheDocument()
  })
})
