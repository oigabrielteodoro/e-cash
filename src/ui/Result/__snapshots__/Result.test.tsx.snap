// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result should be able match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
}

.c0 h1 {
  font-size: 3.2rem;
}

.c0 p {
  margin: 2rem 0 3.2rem;
  font-size: 1.6rem;
}

<div>
  <div
    class="c0"
  >
    <div
      style="width: 180px; height: 180px;"
    >
      <svg
        height="714"
        preserveAspectRatio="xMidYMid meet"
        style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
        viewBox="0 0 796 714"
        width="796"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <clippath
            id="__lottie_element_2"
          >
            <rect
              height="714"
              width="796"
              x="0"
              y="0"
            />
          </clippath>
          <clippath
            id="__lottie_element_13"
          >
            <path
              d="M0,0 L3459,0 L3459,3837 L0,3837z"
            />
          </clippath>
        </defs>
        <g
          clip-path="url(#__lottie_element_2)"
        >
          <g
            clip-path="url(#__lottie_element_13)"
            opacity="1"
            style="display: block;"
            transform="matrix(1,0,0,1,-216,-106)"
          >
            <g
              opacity="0.5"
              style="display: block;"
              transform="matrix(1,0,0,1,1729.5,1918.5)"
            >
              <g
                opacity="1"
                transform="matrix(1.0003999471664429,0,0,0.9998999834060669,0,0)"
              >
                <path
                  d="M0 0"
                  fill="rgb(255,255,255)"
                  fill-opacity="1"
                />
                <path
                  d="M0 0"
                  fill-opacity="0"
                  stroke="rgb(150,218,252)"
                  stroke-linecap="round"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                  stroke-opacity="1"
                  stroke-width="92"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path />
                <path
                  fill-opacity="0"
                  stroke-linecap="round"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path />
                <path
                  fill-opacity="0"
                  stroke-linecap="round"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path />
                <path
                  fill-opacity="0"
                  stroke-linecap="round"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
            <g
              style="display: none;"
            >
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
              <g>
                <path
                  fill-opacity="0"
                  stroke-linecap="butt"
                  stroke-linejoin="miter"
                  stroke-miterlimit="4"
                />
              </g>
            </g>
          </g>
          <g
            style="display: none;"
          >
            <g>
              <path />
              <path
                fill-opacity="0"
                stroke-linecap="butt"
                stroke-linejoin="miter"
                stroke-miterlimit="4"
              />
            </g>
          </g>
          <g
            style="display: none;"
          >
            <g>
              <path />
              <path
                fill-opacity="0"
                stroke-linecap="butt"
                stroke-linejoin="miter"
                stroke-miterlimit="4"
              />
            </g>
          </g>
          <g
            style="display: none;"
          >
            <g>
              <path
                fill-opacity="0"
                stroke-linecap="butt"
                stroke-linejoin="miter"
                stroke-miterlimit="4"
              />
            </g>
          </g>
        </g>
      </svg>
    </div>
    <h1>
      Title
    </h1>
    <p>
      Description
    </p>
    <button>
      Click me
    </button>
  </div>
</div>
`;
